trigger-interpreter/src/iface_impl/trigger.rs

```diff
@@ -45,24 +22,46 @@ impl PubSubTriggerReader {
             authenticator,
         }
     }
+
+    pub fn from_credentials<P: AsRef<Path>>(
+        project_id: String,
+        credentials_file_path: P,
+        subscription: String,
+    ) -> Result<Self> {
+        let authenticator = AuthProvider::from_json_file(credentials_file_path)?;
+        Ok(Self::new(project_id, authenticator, subscription))
+    }
+}
+
+impl Clone for Box<dyn TriggerQueueReader + Send> {
+    fn clone(&self) -> Box<dyn TriggerQueueReader + Send> {
+        self.box_clone()
+    }
 }
 
 impl TriggerQueueReader for PubSubTriggerReader {
-    type Error = Error;
+    fn box_clone(&self) -> Box<dyn TriggerQueueReader + Send> {
```

Comment:
![](https://i.kym-cdn.com/photos/images/facebook/000/764/965/47a.jpg)

Reply:
box_clone is best clone!